pkg/connector/kairosdb.go

```diff
@@ -0,0 +1,494 @@
+	// TODO: make configurable
+	query := MetricsQuery{Start_relative: map[string]string{"value": "3600", "unit": "seconds"}}
+	for _, metric := range JSONmetrics["results"] {
+		var m = map[string]string{"name": metric}
+		query.Metrics = append(query.Metrics, m)
+	}
+	jsonData, err := json.Marshal(query)
+	if err != nil {
+		return fmt.Errorf("kairosdb[%s]: unable to marshal JSON data: %s", connector.name, err)
+	}
+
+	request, err = http.NewRequest("POST", strings.TrimSuffix(connector.URL, "/")+kairosdbURLMetricsTags, bytes.NewBuffer(jsonData))
+	if err != nil {
+		return fmt.Errorf("kairosdb[%s]: unable to set up HTTP request: %s", connector.name, err)
+	}
+
+	request.Header.Add("User-Agent", "Facette")
+	request.Header.Add("X-Requested-With", "KairosdbConnector")
+	request.Header.Set("Content-Type", "application/json")
+
+	response, err = httpClient.Do(request)
+	if err != nil {
+		return fmt.Errorf("kairosdb[%s]: unable to perform HTTP request: %s", connector.name, err)
+	}
+	defer response.Body.Close()
+
+	if err = kairosdbCheckBackendResponse(response); err != nil {
+		return fmt.Errorf("kairosdb[%s]: invalid HTTP backend response: %s", connector.name, err)
+	}
+
+	data, err = ioutil.ReadAll(response.Body)
+	if err != nil {
+		return fmt.Errorf("kairosdb[%s]: unable to read HTTP response body: %s", connector.name, err)
+	}
+	if err = json.Unmarshal(data, &JSONquery); err != nil {
+		return fmt.Errorf("kairosdb[%s]: unable to unmarshal JSON data: %s", connector.name, err)
+	}
+
+	// REVIEW: If we really remove pattern capability above, the code becomes more simply here
+	for _, q := range JSONquery["queries"] {
+		for _, r := range q["results"] {
+			metricName := r.Name
+			m, err := matchRePattern("metric", connector.metricRe, metricName)
+			if err != nil {
+				logger.Log(logger.LevelInfo, "connector", "kairosdb[%s]: metric `%s' does not match pattern, ignoring", connector.name, metricName)
+				continue
+			}
+
+			sources := make(map[string]string)
+			tag := ""
+			for _, t := range connector.srcTags {
+				_, ok := r.Tags[t]
+				if !ok {
+					continue
+				}
+				tag = t
+				for _, sourceName := range r.Tags[t] {
+					s, err := matchRePattern("source", connector.sourceRe, sourceName)
+					if err != nil {
+						logger.Log(logger.LevelInfo, "connector", "kairosdb[%s]: source `%s' does not match pattern, ignoring", connector.name, sourceName)
+						continue
+					}
+					if len(s) > 0 {
+						sources[sourceName] = s
+					} else {
+						sources[sourceName] = sourceName
+					}
+				}
+				break
+			}
+
+			if len(sources) > 0 {
+				mk := metricName
+				if m != metricName && len(m) > 0 {
+					mk = m
+				}
+				for k, v := range sources {
+					sk := k
+					if k != v {
+						sk = v
+					}
+					if _, ok := connector.series[sk]; !ok {
+						connector.series[sk] = make(map[string]kairosdbSeriesEntry)
+					}
+					// TODO: make aggregator configurable via metrics pattern
```

Comment:
Since Facette doesn't do sub-second sampling precision – and I guess it's safe to assume that most use cases won't do either – we can hardcode a 1 second aggregation, and let the aggregation function (min, max, average...) up to the user in the configuration.